internal/util/cmdutil/cmdutil.go

```diff
@@ -194,67 +197,87 @@ func GetKeywordsFromFlag(cmd *cobra.Command) []string {
+		function.TypeMatcher{FnType: cmd.Flag("type").Value.String()},
+		function.KeywordsMatcher{Keywords: GetKeywordsFromFlag(cmd)},
 	}
-	return dedup
+	functions := DiscoverFunctions(cmd)
+	matched := function.MatchFunctions(functions, matchers...)
+	return function.GetNames(matched)
 }
 
 // SuggestKeywords looks for all the unique keywords from Porch functions. This keywords
 // can later help users to select functions.
 func SuggestKeywords(cmd *cobra.Command) []string {
-	functions := porch.MatchFunctions(cmd.Context(), porch.TypeMatcher{FnType: cmd.Flag("type").Value.String()})
-	return porch.UnifyKeywords(functions)
+	functions := DiscoverFunctions(cmd)
+	matched := function.MatchFunctions(functions, function.TypeMatcher{FnType: cmd.Flag("type").Value.String()})
+	return porch.UnifyKeywords(matched)
 }
 
-// FetchFunctionImages returns the list of latest function images from catalog.kpt.dev
-func FetchFunctionImages() []string {
+func DiscoverFunctions(cmd *cobra.Command) []v1alpha1.Function {
+	porchFns := porch.FunctionListGetter{Ctx: cmd.Context()}.Get()
+	catalogV2Fns := fetchCatalogFunctions()
+	return append(porchFns, catalogV2Fns...)
+}
+
+// fetchCatalogFunctions returns the list of latest function images from catalog.kpt.dev.
+func fetchCatalogFunctions() []v1alpha1.Function {
 	content, err := httputil.FetchContent(FunctionsCatalogURL)
 	if err != nil {
 		return nil
 	}
-
-	return listImages(content)
+	return parseFunctions(content)
 }
 
 // fnName -> v<major>.<minor> -> catalogEntry
 type catalogV2 map[string]map[string]struct {
 	LatestPatchVersion string
 	Examples           interface{}
+	Types              []string
+	Keywords           []string
 }
 
 // listImages returns the list of latest images from the input catalog content
-func listImages(content string) []string {
-	var result []string
+func parseFunctions(content string) []v1alpha1.Function {
 	var jsonData catalogV2
 	err := json.Unmarshal([]byte(content), &jsonData)
 	if err != nil {
-		return result
+		return nil
 	}
+	var functions []v1alpha1.Function
+	var fnTypes []v1alpha1.FunctionType
+	var keywords []string
 	for fnName, fnInfo := range jsonData {
 		var latestVersion string
 		for _, catalogEntry := range fnInfo {
 			version := catalogEntry.LatestPatchVersion
 			if semver.Compare(version, latestVersion) == 1 {
 				latestVersion = version
+				keywords = catalogEntry.Keywords
+				for _, tp := range catalogEntry.Types {
+					switch tp {
+					case "validator":
+						fnTypes = append(fnTypes, v1alpha1.FunctionTypeValidator)
+					case "mutator":
+						fnTypes = append(fnTypes, v1alpha1.FunctionTypeMutator)
+					}
+				}
 			}
 		}
-		result = append(result, fmt.Sprintf("%s:%s", fnName, latestVersion))
+		fnName := fmt.Sprintf("%s:%s", fnName, latestVersion)
+		functions = append(functions,
+			v1alpha1.Function{
+				TypeMeta: metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      fnName,
+					Namespace: function.CatalogV2,
```

Comment:
What's this? Using `__catalog_v2` as namespace?

Reply:
This is an invalid namespace used to distinguish catalog function from porch functions. This will be trimmed before showing to users.
What users will see is something like: "default" is the porch function namespace. `set-namespace:v0.3.3` comes from catalog
```bash
 kpt fn eval yu --keywords namespace -i [tab tab]
default:kpt-functions:set-annotation:v0.1-yuwenma  default:kpt-functions:set-namespace:v0.3-yuwenma
default:kpt-functions:set-annotation:v0.2-yuwenma  set-namespace:v0.3.3
default:kpt-functions:set-annotation:v0.3-yuwenma
```

Reply:
we could consider using empty value for this placeholder, or (worst case) using some wrapper object for the Function resource(s).

that said, I think if documented for posterity, this approach seems fine to me.

Reply:
Sg. I used wrapper and added doc string.